roles/branch-validation/tasks/ha.yml

```diff
@@ -1,114 +1,105 @@
 ---
-## HA testing
+- name: begin ha.yml
+  debug:
+    msg: Running ha.yml playbook
+
+- name: remove previously allocated floating ip
+  os_floating_ip:
+    auth: "{{ auth }}"
+    network: external
+    server: "validate-{{ groups[instance_group][0] }}"
+    floating_ip_address: "{{ fip.floating_ip.floating_ip_address }}"
+    state: absent
+  failed_when: false
+
 - name: add floating IPs to instances on controllers
   os_floating_ip:
-    auth:
-      username: admin
-      project_name: admin
-      password: "{{ secrets.admin_password }}"
-      auth_url: "http://127.0.0.1:{{ endpoints.keystone_admin.port.backend_api }}/{{ endpoints.keystone.version }}"
+    auth: "{{ auth }}"
     network: external
     server: "validate-{{ item }}"
     reuse: true
   register: fip_controllers
   with_items: groups[ha_instance_group]
+  when: converged
+
+- name: floating ips are reachable
+  command: "ping -c 5 {{ item.floating_ip.floating_ip_address }}"
+  with_items: fip_controllers.results
+  register: pings
+  until: pings|success
+  delay: 1
+  retries: 5
+  sudo: no
+  delegate_to: localhost
+  when: converged
 
-- name: add floating ip to a non-controller instance
+- name: add floating IP to a non-controller instance
   os_floating_ip:
-    auth:
-      username: admin
-      project_name: admin
-      password: "{{ secrets.admin_password }}"
-      auth_url: "http://127.0.0.1:{{ endpoints.keystone_admin.port.backend_api }}/{{ endpoints.keystone.version }}"
+    auth: "{{ auth }}"
     network: external
     server: "validate-{{ item }}"
     reuse: true
   register: fip_compute
+  ignore_errors: yes
   with_items:
-    - "{{ (groups['compute']|difference(groups[ha_instance_group]))[0] }}"
+    - "{{ (groups[instance_group][1:]|difference(groups[ha_instance_group]))[0] }}"
+  when: not converged
 
 - name: floating ips are reachable
   command: "ping -c 5 {{ item.floating_ip.floating_ip_address }}"
-  with_items: fip_controllers.results + fip_compute.results
-  changed_when: false
+  with_items: fip_compute.results
   register: pings
   until: pings|success
-  delay: 1
+  delay: 10
   retries: 5
+  sudo: no
   delegate_to: localhost
+  when: not converged
 
-- name: define ha master
-  set_fact: master=true
+- name: define master
+  set_fact:
+    master: true
   when: floating_ip in ansible_all_ipv4_addresses
 
+- name: define secondary
+  set_fact:
+    master: false
+  when: floating_ip not in ansible_all_ipv4_addresses
+
 # initiate hard reboot on the current master controller
-- name: initiate a failover on controller 0
-  shell: echo 'b' | tee /proc/sysrq-trigger
-  when: master is defined
+- name: reboot controller 0 (master)
+  shell: shutdown -r now
```

Comment:
I wouldn't call this a hard reboot.  We have seen in the past that HA will accommodate a graceful shutdown (like above) but will fail when a true panic scenario occurs.  So if we really want hard reboot we should do something like this: https://major.io/2009/01/29/linux-emergency-reboot-or-shutdown-with-magic-commands/